src/Build/BackEnd/Components/RequestBuilder/TargetUpToDateChecker.cs

```diff
@@ -1075,6 +1085,175 @@ internal static bool IsAnyOutOfDate<T>(out DependencyAnalysisLogDetail dependenc
+            //         possibly the outputs list isn't actually the shortest list. However it always is the shortest
+            //         in the cases I've seen, and adding this optimization would make the code hard to read.
+            (string filePath, DateTime fileTime) oldestOutput = GetOldestFile(projectDirectory, outputs);
+            (string filePath, DateTime fileTime) newerInput = GetNewerFile(projectDirectory, inputs, oldestOutput.fileTime);
+            dependencyAnalysisDetailEntry = CompareInputAndOutput(newerInput, oldestOutput);
+            return dependencyAnalysisDetailEntry != null;
+        }
+
+        private static void ValidateTargetItemsType<T>(IList<T> items)
+        {
+            if (items.Count > 0)
+            {
+                ErrorUtilities.VerifyThrow(items[0] is string || items[0] is ProjectItemInstance, "Must be either string or ProjectItemInstance");
+            }
+        }
+
+        private static DependencyAnalysisLogDetail CompareInputAndOutput((string filePath, DateTime fileTime) newerInput, (string filePath, DateTime fileTime) oldestOutput)
+        {
+            bool isMissingOutput = IsInvalidFileTime(oldestOutput.fileTime);
+            bool isMissingInput = IsInvalidFileTime(newerInput.fileTime);
+            bool isTargetUpToDate = !isMissingOutput && !isMissingInput && newerInput.fileTime <= oldestOutput.fileTime;
+
+            // All exist and no inputs are newer than any outputs; up to date
+            if (isTargetUpToDate)
+            {
+                return null;
+            }
+
+            OutofdateReason reason = isMissingOutput ? OutofdateReason.MissingOutput
+                : isMissingInput ? OutofdateReason.MissingInput : OutofdateReason.NewerInput;
+            return new DependencyAnalysisLogDetail(newerInput.filePath, oldestOutput.filePath, null, null, reason);
+        }
+
+        private static bool IsInvalidFileTime(DateTime lastWriteTimeUtc)
+        {
+            return lastWriteTimeUtc == DateTime.MinValue;
+        }
+
+        private static (string filePath, DateTime fileTime) GetOldestFile<T>(string projectDirectory, IList<T> escapedFilePaths)
+        {
+            string oldestFilePath = String.Empty;
+            DateTime oldestFileTime = DateTime.MaxValue;
+            IDictionary<string, ISet<string>> directoryToFileNames = MapDirectoryToFileNames(escapedFilePaths);
+
+            foreach (string directoryName in directoryToFileNames.Keys)
+            {
+                ISet<string> fileNamesToFind = directoryToFileNames[directoryName];
+                IEnumerable<(string fileName, DateTime fileTime)> files = GetFilesInDirectory(projectDirectory, directoryName);
+
+                foreach ((string fileName, DateTime fileTime) in files)
+                {
+                    bool isOlderFile = fileNamesToFind.Remove(fileName) && fileTime < oldestFileTime;
```

Comment:
I think checking the set if the filename is there is cheaper than removing the file from the set

Reply:
Ignore that, apparently Remove is o(1) :)

---
In reply to: [210367450](https://github.com/Microsoft/msbuild/pull/3608#discussion_r210367450) [](ancestors = 210367450)